File: probability-example-main/probability-example-PythonVersion/python/Deck.py
import enum
import itertools
import Card
import random

class Comparator:
    LESS = -1
    EQUAL = 0
    GREATER = 1

class Deck:
    SUIT_LIST = ['Spade', 'Heart', 'Club', 'Diamond'];
    VALUE_LIST = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'K', 'Q', 'K', 'A']
    def __init__(self):
        self.cards = [Card.Card(x[0], x[1]) for x in itertools.product(Deck.SUIT_LIST, Deck.VALUE_LIST)]
        self.discard = []

    def shuffle_all(self):
        self.cards = self.cards + self.discard
        self.discard = []
        random.shuffle(self.cards)

    def shuffle_current(self):
        random.shuffle(self.cards)

    def draw_and_discard(self):
        card = self.cards.pop()
        self.discard.append(card)
        return card # Beware, this is a reference

    def draw_and_replace(self):
        self.cards.insert(0, self.cards.pop())
        return self.cards[0] # Beware, this is a reference

    def count(self):
        return len(self.cards)

    def compare(card1, card2):
        if Deck.VALUE_LIST.index(card1.value) < Deck.VALUE_LIST.index(card2.value):
            return Comparator.LESS
        elif Deck.VALUE_LIST.index(card1.value) > Deck.VALUE_LIST.index(card2.value):
            return Comparator.GREATER
        else:
            return Comparator.EQUAL

if __name__ == "__main__":
    deck = Deck()
    deck.shuffle_all()
    print(deck.draw_and_replace().display())
    print('Size of deck:', len(deck.cards))
    print(deck.draw_and_discard().display())
    print('Size of deck:', len(deck.cards))
    print('Size of discard:', len(deck.discard))
    deck.shuffle_current()
    print('Size of deck:', len(deck.cards))
    print('Size of discard:', len(deck.discard))
    deck.shuffle_all()
    print('Size of deck:', len(deck.cards))
    print('Size of discard:', len(deck.discard))
    card1 = deck.draw_and_replace()
    card2 = deck.draw_and_replace()

    # Got a duplicate here once
    print(card1.display(), card2.display(), Deck.compare(card1, card2))

File: probability-example-main/probability-example-PythonVersion/csharp/Deck.cs
using System;

public class Deck{

    public static readonly string []SUIT_ARRAY = {"Spade","Heart","Club","Diamond"};
    public static readonly string []VALUE_ARRAY = {"A","2","3","4","5","6","7","8","9","10","J","Q","K"};
    private static Random rnd = new Random();

    private List<Card> _deck=new List<Card>();
    private List<Card> _discard=new List<Card>();

    public Deck(){
        foreach (var s in SUIT_ARRAY){
            foreach (var v in VALUE_ARRAY){
                _deck.Add(new Card(s,v));
            }
        }
        ShuffleAll();
    }

    //Implents the Fisher Yates algorithm
    //Shuffles back in the discard pile
    public void ShuffleAll(){
        if(_discard.Count > 0){
            foreach(Card t in _discard){
                _deck.Add(t);
            }
            _discard.Clear();
        }
        for (int i = _deck.Count-1; i > 0; i--){
            int randomIndex = rnd.Next(0, i + 1);
        
            var temp = _deck[i];
            _deck[i] = _deck[randomIndex];
            _deck[randomIndex] = temp;
        }
    }

    //Implements Fisher Yates
    //Shuffles only the current stack, not the discard
    public void ShuffleCurrent(){
        for (int i = _deck.Count-1; i > 0; i--){
            int randomIndex = rnd.Next(0, i + 1);
        
            var temp = _deck[i];
            _deck[i] = _deck[randomIndex];
            _deck[randomIndex] = temp;
        }
    }


    //returns the first card in the deck, removes it and puts it in _discard
    public Card DrawAndDiscard(){
        Card temp = _deck[0];
        _deck.RemoveAt(0);
        _discard.Add(temp);
        return temp;
    }

    //returns the first card in the deck and reshuffles
    public Card DrawAndReplace(){
        Card temp = _deck[0];
        ShuffleAll();
        return temp;
    }

    public int Count(){
        return _deck.Count;
    }

    //Compares two cards
    //TODO Deal with suits.
    public static Comparator Compare (Card x, Card y){
        Comparator temp;
        int xIndex = Array.IndexOf(VALUE_ARRAY, x.GetValue()); 
        int yIndex = Array.IndexOf(VALUE_ARRAY, y.GetValue()); 
        if( xIndex < yIndex){
            temp= Comparator.Less;
        }
        else if( xIndex > yIndex){
            temp= Comparator.Greater;
        }
        else{
            temp =  Comparator.Equal;
        }
        return temp;
    }
}